core/new-gui/src/app/workspace/service/dynamic-schema/input-schema-propagation/input-schema-propagation.service.ts

```diff
@@ -0,0 +1,163 @@
+      .subscribe(({ operatorID, oldInputSchema, newInputSchema }) => {
+        this.updateOperatorPropertiesOnInputSchemaChange(operatorID, oldInputSchema, newInputSchema);
+      });
+  }
+
+  private updateOperatorPropertiesOnInputSchemaChange(
+    operatorID: string,
+    oldInputSchema: OperatorInputSchema,
+    newInputSchema: OperatorInputSchema
+  ): void {
+    const dynamicSchema = this.dynamicSchemaService.getDynamicSchema(operatorID);
+    // console.log(dynamicSchema, oldInputSchema, newInputSchema);
+    if (!dynamicSchema.jsonSchema.properties) return;
+    if (!oldInputSchema) return;
+    const operator = this.workflowActionService.getTexeraGraph().getOperator(operatorID);
+    // console.log(operator)
+    for (let port = 0; port < oldInputSchema.length; ++port) {
+      // TODO: the substitute cost function can be improved to reach expected behaviour
+      const attributeMapping = levenshtein(
+        oldInputSchema[port],
+        newInputSchema[port],
+        (_: any) => 1,
+        (_: any) => 1,
+        (a: SchemaAttribute, b: SchemaAttribute) => {
+          if (a.attributeName === b.attributeName && a.attributeType === b.attributeType) return 0;
+          else if (a.attributeName !== a.attributeName && a.attributeType !== b.attributeType) return 2;
+          else return 1;
+        }
+      );
+      // [[oldAttribute1, newAttribute1], [oldAttribute2, newAttribute2], ...]
+      const mapping: SchemaAttribute[][] = attributeMapping.pairs();
+      // Resolve #ref in some json schema (e.g. Filter)
+      const resolvedJsonSchema = jsonRefLite.resolve(dynamicSchema.jsonSchema);
+      // console.log("resolved", resolvedJsonSchema)
+      const newProperty = this.updateOperatorProperty(
+        operator.operatorProperties,
+        mapping,
+        resolvedJsonSchema.properties,
+        port
+      );
+      this.workflowActionService.setOperatorProperty(operatorID, newProperty);
+      // console.log("updated", this.workflowActionService.getTexeraGraph().getOperator(operatorID));
+    }
+  }
+
+  private updateOperatorProperty(
+    currentProperties: Readonly<{ [key: string]: any }>,
+    mapping: SchemaAttribute[][],
+    jsonSchemaProperties: { [key: string]: boolean | CustomJSONSchema7 },
+    port: number
+  ) {
+    // console.log("properties: ", currentProperties);
+    // console.log("mapping: ", mapping);
+    const updatePropertyRecurse = (
+      properties: any,
+      jsonSchema: { [key: string]: CustomJSONSchema7 | boolean } | undefined
+    ): boolean => {
+      // console.log("recurse:", properties, jsonSchema)
+      if (!jsonSchema) {
+        return false;
+      }
+      for (const key in jsonSchema) {
+        const jsonSchemaProperty = jsonSchema[key];
+        if (typeof jsonSchemaProperty === "boolean") {
+          continue;
+        }
+        if (jsonSchemaProperty.autofillAttributeOnPort === port) {
+          if (jsonSchemaProperty.type === "array") {
```

Comment:
from what I understand: `updatePropertyRecurse` walks through the whole tree,  lines 110 to 136 handle the property changes.

Can you extract the code in  lines 110 to 136 to two separate functions? So it's more clear that we separate the recursive walk framework, and the property change handling logic.